[RFC]uart: Add support for RS485

kumarnac opened the pull request:
Adds support for configuring and enabling RS485 signalling
on supported uarts.

Signed-off-by: Nachiketa Kumar <nachiketa.kumar@intel.com>

kumarnac committed:
uart: Add support for RS485

Adds support for configuring and enabling RS485 signalling
on supported uarts.

Signed-off-by: Nachiketa Kumar <nachiketa.kumar@intel.com>

talih0 commented on drivers/serial/Kconfig:
```diff
@@ -31,6 +31,12 @@ config SERIAL_SUPPORT_INTERRUPT
+	bool
+	help
+	  This is an option to be enabled by individual serival driver
+	  to signal that the driver and hardware supports RS_455 interface.
```
typo /RS_455/RS_485

kumarnac replied:
Thanks for pointing  out the typo.  Fixed in the latest revision

dcpleung commented on include/drivers/uart.h:
```diff
@@ -237,6 +241,57 @@ struct uart_event {
+
+} uart_rs485_transfer_mode_t;
+
+/**$
```
Remove extra '$' sign.

dcpleung reviewed (commented):
Just wondering how complicated it would be if you roll this into the existing `uart_configure()`.

nashif commented on include/drivers/uart.h:
```diff
@@ -237,6 +239,55 @@ struct uart_event {
+
+	/** Polarity for RS485 - Active High  **/
+	UART_RS485_POL_ACTIVE_HIGH
+} uart_rs485_polarity_t;
```
here you call it `rs485`...

kumarnac replied:
Changed in latest vesrion

nashif commented on include/drivers/uart.h:
```diff
@@ -237,6 +239,55 @@ struct uart_event {
+	UART_RS485_POL_ACTIVE_HIGH
+} uart_rs485_polarity_t;
+
+struct uart_rs_485_config {
```
and here you call it `_rs_485_`. Can you please use one convention? (rs486 without underscore maybe)

kumarnac replied:
Thank for reviewing Anas,  will make the naming uniform.

kumarnac commented:
Detail on RS485 signalling requirements can be found at https://en.wikipedia.org/wiki/RS-485
This change provides enabling and configuration options for different parameters of RS-485

anthonytex commented:
May i suggest to include also a general binding like the one present in the Linux kernel to allow for example setting the 485 mode at compile time?
Thanks 
 https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml

galak commented:
Is the a driver side or new driver that would utilize RS485 that you can also post?

galak commented:
want to see how the params here would work in a simpler UART case that uses RTS to the transciever.

galak commented:
Need to see how we scale the UART api to multiple instances but ones that do an don't support the feature.  (This is a general problem that also exists for async support).

carlescufi commented:
API meeting:

- Would be great to pull more information from Devicetree
- Does the Kconfig option enable this functionality for all UART instances?
  - This applies to the async option as well, so should be fixed for both

kumarnac commented:
> Is the a driver side or new driver that would utilize RS485 that you can also post?
The RS485 driver would be a part of a new board with dependencies on external  HAL module and thus not added to this change. 

> Need to see how we scale the UART api to multiple instances but ones that do an don't support the feature. (This is a general problem that also exists for async support).
Explored more on this  and   found that it would be problem for instances that do not support  RS485.




> API meeting:
> 
> * Would be great to pull more information from Devicetree
> * Does the Kconfig option enable this functionality for all UART instances?
>   
>   * This applies to the async option as well, so should be fixed for both

Modified the pull request to include rs485 configuration from the device tree and removed the changes introduced in headers. 
On enabling CONFIG_SERIAL_SUPPORT_RS485 driver implementations can use the uart-rs485-controller.yaml  options to enable instance specific configurations.  Enabling functions would be private implementations let to drivers.

kumarnac commented:
> May i suggest to include also a general binding like the one present in the Linux kernel to allow for example setting the 485 mode at compile time?
> Thanks
> https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml

Made change to add uart-rs485-controller.yaml . Can you please take a look.
Thanks

alexanderwachter commented:
API meeting:

bindings are looking good. Please revisit the types and see whether you can find better ones, like bool to turn on/off something.
@kumarnac mentioned that the Linux bindings are incomplete, so we cannot use them as they are.
@galak, can you check against the Linux driver and give your acknowledgment for not following it exactly?

anthonytex reviewed (commented):
These are my suggestions.

carlescufi commented:
@kumarnac will you pick up this PR again to complete the work?
@anthonytex failing the above, would you consider picking it up yourself?

anthonytex commented:
Hello @carlescufi , Yes i'm going to continue the work if this pr will be closed thanks

anthonytex commented:
@carlescufi i've opened a new PR, mentioning this please see https://github.com/zephyrproject-rtos/zephyr/pull/40907
